File: javascript/src/Literal.ts
import { Value } from './value';

export class Literal<T> extends Value {
	protected data: T;

	constructor(data: T) {
		super();

		this.data = data;
	}

	toString(): string {
		return String(this.data);
	}

	
	run() {
		return this;
	}

	toInt(): number {
		return Number(this.data);
	}

	toBool(): boolean {
		return Boolean(this.data);
	}

	eql(rhs: any): boolean {
		let z: any = rhs;
		return rhs instanceof this.constructor && (this.data === z.data);
	}
}

File: csharp/Literal.cs
using System;

namespace Knight
{
	public abstract class Literal<T> : IValue, IEquatable<IValue>
	where T: IEquatable<T>
	{
		protected readonly T _data;

		public Literal(T data) => _data = data;
		public IValue Run() => this;
		public abstract void Dump();

		public override string ToString() => _data.ToString();
		public abstract bool ToBoolean();
		public abstract long ToNumber();

		public bool Equals(IValue obj) => GetType() == obj.GetType() && _data.Equals(((Literal<T>) obj)._data);
	}
}